cmake: New target which generates a sort of development kit for llext (#69831)
zephyrproject-rtos/zephyr: samples/subsys/llext/edk/README.rst

> .. zephyr:code-sample:: llext-edk
>    :name: Linkable loadable extensions EDK
> 

Reviewer: ```suggestion
.. zephyr:code-sample:: llext-edk
   :name: Linkable loadable extensions EDK
   :relevant-api: llext

   Enable linkable loadable extension development outside the Zephyr tree using LLEXT EDK (Extension Development Kit)
```

About
*****

This sample demonstrates how to use the Zephyr LLEXT EDK (Extension Development
Kit). It is composed of one Zephyr application, which provides APIs for the

> extensions that it loads. The API provided is a simple publish/subscribe system,

Reviewer: ```suggestion
extensions that it loads. The provided API is a simple publish/subscribe system,
```

> based on Zbus, which extensions use to communicate with each other.

Reviewer: ```suggestion
based on :ref:`Zbus <zbus>`, which extensions use to communicate with each other.
```

The application is composed of a subscriber thread, which listens for events
published and republishes then via Zbus to the extensions that are
subscribers. There are four extensions, which are loaded by the application and
run in different contexts. Extensions ``ext1``, ``ext2`` and ``ext3`` run in
userspace, each demonstrating different application and Zephyr API usage, such as
semaphores, spawning threads to listen for events or simply publishing or
subscribing to events. Extension ``kext1`` runs in a kernel thread, albeit similar
to ``ext3``.

The application also creates different memory domains for each extension, thus
providing some level of isolation - although the kernel one still has access
to all of Zephyr kernel.

> Note that the kernel extension is only available when the EDK is built with
> ``CONFIG_LLEXT_EDK_USERSPACE_ONLY=n``.

Reviewer: ```suggestion
Note that the kernel extension is only available when the EDK is built with the
:kconfig:`CONFIG_LLEXT_EDK_USERSPACE_ONLY` option disabled.
```

during build time, which is not really practical. This sample is about the EDK
providing the ability to build extensions independently from the application.
One could build the extensions in different directories, not related to the
Zephyr application - even on different machines, using only the EDK. At the
limit, one could even imagine a scenario where the extensions are built by
different teams, using the EDK provided by the application developer.

Building the EDK
****************

To build the EDK, use the ``llext-edk`` target. For example:

.. zephyr-app-commands::
   :zephyr-app: samples/subsys/llext/edk/app
   :board: qemu_cortex_r5
   :goals: build llext-edk
   :west-args: -p=always
   :compact:

Copy the EDK to some place and extract it:

.. code-block:: console

    mkdir /tmp/edk
    cp build/zephyr/llext-edk.tar.xz /tmp/edk
    cd /tmp/edk
    tar -xf llext-edk.tar.xz

Then set ``LLEXT_EDK_INSTALL_DIR`` to the extracted directory:

.. code-block:: console

    export LLEXT_EDK_INSTALL_DIR=/tmp/edk/llext-edk

This variable is used by the extensions to find the EDK.

Building the extensions
***********************

> First, ensure the Zephyr SDK location is available at the environment variable
> ``ZEPHYR_SDK_INSTALL_DIR``:
> 
> .. code-block:: console
> 
>     export ZEPHYR_SDK_INSTALL_DIR=</path/to/zephyr-sdk>
> 
> As the ``ZEPHYR_SDK_INSTALL_DIR`` is used by the extensions to find the Zephyr SDK.
> 

Reviewer: ```suggestion
The :envvar:`ZEPHYR_SDK_INSTALL_DIR` environment variable is used by the
extensions to find the Zephyr SDK, so you need to ensure it's properly set:

.. code-block:: console

    export ZEPHYR_SDK_INSTALL_DIR=</path/to/zephyr-sdk>

```

The application is built using the Zephyr build system. The EDK is built using
the Zephyr build system as well, via ``llext-edk`` target. The EDK is then
extracted and the extensions are built using CMake.

Finally, the way the application loads the extensions is by including them
during build time, which is not really practical. This sample is about the EDK
providing the ability to build extensions independently from the application.
One could build the extensions in different directories, not related to the
Zephyr application - even on different machines, using only the EDK. At the
limit, one could even imagine a scenario where the extensions are built by
different teams, using the EDK provided by the application developer.

Building the EDK
****************

To build the EDK, use the ``llext-edk`` target. For example:

.. zephyr-app-commands::
   :zephyr-app: samples/subsys/llext/edk/app
   :board: qemu_cortex_r5
   :goals: build llext-edk
   :west-args: -p=always
   :gen-args: -DSDK_BUILD=1
   :compact:

Note that the ``SDK_BUILD`` option is necessary to build the EDK initially,
as it prevents the application from trying to load the extensions before

> they are built.

Author: But if I build the EDK from another app, it won't include the application specific syscalls. Agree `SDK_BUILD` is confusing, will see if I can avoid it somehow.

.. zephyr:code-sample:: llext-edk
   :name: Linkable loadable extensions EDK
   :relevant-api: llext

Enable linkable loadable extension development outside the Zephyr tree using

> LLEXT EDK (Extension Development Kit).

Reviewer: Needs 3-space indent as this belongs to the code-sample directive

Zephyr application - even on different machines, using only the EDK. At the
limit, one could even imagine a scenario where the extensions are built by
different teams, using the EDK provided by the application developer.

Building the EDK
****************

To build the EDK, use the ``llext-edk`` target. For example:

.. zephyr-app-commands::
   :zephyr-app: samples/subsys/llext/edk/app
   :board: qemu_cortex_r5
   :goals: build llext-edk
   :west-args: -p=always
   :compact:

Copy the EDK to some place and extract it:

.. code-block:: console

    mkdir /tmp/edk
    cp build/zephyr/llext-edk.tar.xz /tmp/edk
    cd /tmp/edk
    tar -xf llext-edk.tar.xz

Then set ``LLEXT_EDK_INSTALL_DIR`` to the extracted directory:

.. code-block:: console

    export LLEXT_EDK_INSTALL_DIR=/tmp/edk/llext-edk

This variable is used by the extensions to find the EDK.

Building the extensions
***********************

The :envvar:`ZEPHYR_SDK_INSTALL_DIR` environment variable is used by the
extensions to find the Zephyr SDK, so you need to ensure it's properly set:

.. code-block:: console

    export ZEPHYR_SDK_INSTALL_DIR=</path/to/zephyr-sdk>

> 
> As the ``ZEPHYR_SDK_INSTALL_DIR`` is used by the extensions to find the Zephyr SDK.

Reviewer: ```suggestion
```
